src/lib.rs

```diff
@@ -0,0 +1,210 @@
+            readiness_check_port: env::var("READINESS_CHECK_PORT")
+                .unwrap_or_else(|_| env::var("PORT").unwrap_or_else(|_| "8080".to_string())),
+            readiness_check_path: env::var("READINESS_CHECK_PATH").unwrap_or_else(|_| "/".to_string()),
+            base_path: env::var("REMOVE_BASE_PATH").ok(),
+            async_init: env::var("ASYNC_INIT")
+                .unwrap_or_else(|_| "false".to_string())
+                .parse()
+                .unwrap_or(false),
+        }
+    }
+}
+
+pub struct Adapter {
+    client: Client,
+    healthcheck_url: String,
+    async_init: bool,
+    ready_at_init: bool,
+    domain: Url,
+    base_path: Option<String>,
+}
+
+impl Adapter {
+    /// Create a new Adapter instance.
+    /// This function initializes a new HTTP client
+    /// to talk with the web server.
+    pub fn new(options: &AdapterOptions) -> Adapter {
+        let client = Client::builder()
+            .redirect(redirect::Policy::none())
+            .pool_idle_timeout(Duration::from_secs(4))
+            .build()
+            .unwrap();
+
+        let healthcheck_url = format!(
+            "http://{}:{}{}",
+            options.host, options.readiness_check_port, options.readiness_check_path
+        );
+
+        let domain = format!("http://{}:{}", options.host, options.port).parse().unwrap();
+
+        Adapter {
+            client,
+            healthcheck_url,
+            domain,
+            base_path: options.base_path.clone(),
+            async_init: options.async_init,
+            ready_at_init: false,
+        }
+    }
+
+    /// Switch the default HTTP client with a different one.
+    pub fn with_client(self, client: Client) -> Self {
+        Adapter { client, ..self }
+    }
+
+    /// Check if the web server has been initialized.
+    /// If `Adapter.async_init` is true, cancel this check before
+    /// Lambda's init 10s timeout, and let the server boot in the background.
+    pub async fn check_init_health(&mut self) {
+        self.ready_at_init = if self.async_init {
+            timeout(Duration::from_secs_f32(9.8), self.check_readiness())
+                .await
+                .unwrap_or_default()
+        } else {
+            self.check_readiness().await
+        };
+    }
+
+    async fn check_readiness(&self) -> bool {
+        Retry::spawn(FixedInterval::from_millis(10), || {
+            let fut = self.check_health_url();
+            async move { fut.await }
+        })
+        .await
+        .unwrap()
+    }
+
+    async fn check_health_url(&self) -> Result<bool, std::convert::Infallible> {
+        self.client
+            .head(&self.healthcheck_url)
+            .send()
+            .await
+            .map(|r| r.status().is_success())
+            .or(Ok(false))
+    }
+}
+
+/// Implement a `Tower.Service` that sends the requests
+/// to the web server.
+impl Service<Request> for Adapter {
+    type Error = Error;
```

Comment:
Clion highlights this as "Different impl member order from the trait."  Better to swap these two lines.